### DISCUSSION TITLE
Middleware with geolocation

### DISCUSSION THREAD

**Discussion by minnyww** [vercel/next.js#39619] [Help] [answered]
I have a question about the nextjs page middleware, [see here](https://nextjs.org/docs/advanced-features/middleware), according to the documentation the request object should contain geo, IP, ua, cookies and nexturl properties.

So all I am trying to do is get the users location through page middleware like this:

```js
import { NextRequest, NextResponse } from 'next/server';

export async function middleware(req) {
  const { nextUrl, geo } = req;
  console.log(geo) //=> always empty {}
  const country = geo.country || 'US'
  nextUrl.searchParams.set('locale', country);


  return NextResponse.rewrite(nextUrl);
}
 ```
 
 But when I run my project locally the geolocation is an empty object, why?

Will all hosting providers send geolocation data, or does it depend on the provider?

**Comment by gaston-flores** [comment]
No, it depends in the provider. The geolocation info will always be empty in your local environment unfortunately, since next's server has no way of knowing where you are. If you deploy your code to Vercel however you should see the GeoIP data.

**Reply by minnyww** [reply]
How  debug the geolocation in local development ?

**Reply by robertocandales** [reply]
you have to use getServerSideProps in /pages/index.tsx

```
export const getServerSideProps = ({ query }) => ({
  props: query,
});
```

and then you can see the props in index file

```
export default function Index(props) {
console.log(props)
...
}
```

**Reply by suneelx** [reply]
> No, it depends in the provider. The geolocation info will always be empty in your local environment unfortunately, since next's server has no way of knowing where you are. If you deploy your code to Vercel however you should see the GeoIP data.

is there a easy way with aws ec2 or lightsail providing geo? My app is deployed in aws. I tried It seem they don't provide geo location.

**Reply by JeromeIllgner** [reply]
Is there a way of finding out which providers pass the geo headers? Is there any way of providing these headers other than a proxy server outside of vercel?

**Comment by leerob** [comment, marked as answer]
Updated the docs here to make it more clear this depends on the infra provider, but you can also read some things from headers if provided: https://github.com/vercel/next.js/pull/59719
